### PR TITLE
PLT-3466 E10: Add announcement bar feature

### PR DESCRIPTION
**Pull request by jwilander**
#### Summary
Converted the error bar to be a general announcement bar. Added a color setting component to the system console.

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-3466

#### Checklist
- [x] Has UI changes
- [x] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/platform/blob/master/i18n/en.json) and [.../webapp/i18n/en.json](https://github.com/mattermost/platform/tree/master/webapp/i18n/en.json)) updates

### CODE REVIEW

**Comment by jwilander** [comment]
Two UI issues I need @asaadmahmood's help with:

1. The color settings in `system console -> policy` don't show the color in the boxes properly https://github.com/mattermost/platform/pull/6509/files#diff-8c8816b8c6c71a254a6a0df06f4e5a8dR95
2. When `AllowBannerDismissal` is set to false, the banner in the system console is under the sidebar header and adds a second scroll bar to the right

**Comment by jwilander** [comment]
I still need to add some comments to the new component I added and add the disabling of fields in the system console

**Comment by lfbrock** [comment]
Okay, here's some questions & small bugs so far: 
1. Can we switch the default for "Allow banner dismissal" to true? 
2. Can we show the announcement banner only to logged in users? 
3. The enable/disable setting for the banner should disable the other banner related settings when disabled (like the custom branding one does) 
  - screenshot 2017-05-30 11 30 21
4. On old mobile apps, the banner pushes the text box + channel info pages slightly too far off page
  

5. Was planning on having a max length on the text, but testing it out I don't think we need it - so we can take out the "Maximum 100 characters allowed." in the help text (or I can submit a PR to take it out later) 


Formatting / UI: (maybe for @asaadmahmood) 
1. Can we format the permanent banner so it's thinner on desktop? 
2. Can we decrease the width of the color pickers? Color picker is showing up offscreen for me
screenshot 2017-05-30 12 56 11

**Comment by jwilander** [comment]
1. Yep, will do
2. Yes, I can do that
3. Commented about this before, still need to add it
4. This is a CSS issue, can we get @asaadmahmood to look at that?
5. I can update it

**Comment by lfbrock** [comment]
Also I'm not sure if this is related to your changes, but I can't repro on the ci servers - the link color in site URL banner is the incorrect color: 
screenshot 2017-05-30 11 27 51

**Comment by jwilander** [comment]
Yeah I think that's related. I removed an `!important` CSS tag from link colors because it was preventing me from setting them manually. @asaadmahmood would you be able to help with that as well? I'd probably just it up even more

**Comment by jwilander** [comment]
@lfbrock fixed 1,2,3,5

File: model/license.go [hunk 1]
```diff
@@ -142,6 +143,11 @@ func (f *Features) SetDefaults() {
 		f.ElasticSearch = new(bool)
 		*f.ElasticSearch = *f.FutureFeatures
 	}
+
+	if f.Announcement == nil {
+		f.Announcement = new(bool)
+		*f.Announcement = true
```
Comment (hmhealey) [inline model/license.go]: Should this be set to `*f.FutureFeatures`, or is that not for E10 features?

File: webapp/components/announcement_bar/announcement_bar.jsx [hunk 2]
```diff
@@ -115,28 +143,57 @@ export default class ErrorBar extends React.Component {
...
+                    href='#'
+                    className='error-bar__close'
+                    style={linkStyle}
+                    onClick={this.handleClose}
+                >
+                    {'×'}
+                </a>
+            );
+        } else {
+            barStyle.position = 'static';
```
Comment (hmhealey) [inline webapp/components/announcement_bar/announcement_bar.jsx]: Isn't `position: static` the default? If this is about overriding the style applied by some class, maybe we could have two classes like `error-bar--temporary` and `error-bar--permanent` based of whether or not `allowDismissal` is set?

**Comment by asaadmahmood** [comment]
Did those changes, but for the scrolling issue, I would need a class applied to the body tag if the banner is not dismissible. @jwilander

File: webapp/components/announcement_bar/announcement_bar.jsx (hunk 2 above)
Comment (jwilander) [inline webapp/components/announcement_bar/announcement_bar.jsx]: This was temporary until @asaadmahmood made his changes. We should probably wait to dev until we're done with PM review

File: model/license.go (hunk 1 above)
Comment (jwilander) [inline model/license.go]: `*f.FutureFeatures` is used for E20, E10 should be set to true here

**Comment by jwilander** [comment]
@asaadmahmood just replace the `asaadclassname` with your class in the four lines I added it. I also rebase so you might need to clean pull the branch

**Comment by jwilander** [comment]
@lfbrock should be ready for you

File: webapp/components/admin_console/color_setting.jsx [hunk 3]
```diff
@@ -0,0 +1,119 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+import Setting from './setting.jsx';
+
+import React from 'react';
+import PropTypes from 'prop-types';
+import {ChromePicker} from 'react-color';
+
+export default class ColorSetting extends React.PureComponent {
```
Comment (crspeller) [inline webapp/components/admin_console/color_setting.jsx]: Is there not already a colour picker for the theme? Shouldn't they be using the same one?

Comment (jwilander) [inline webapp/components/admin_console/color_setting.jsx]: That one uses jquery all over the place and isn't very React friendly. We should switch those ones to be like this

Comment (crspeller) [inline webapp/components/admin_console/color_setting.jsx]: Ticket?

Comment (jwilander) [inline webapp/components/admin_console/color_setting.jsx]: https://mattermost.atlassian.net/browse/PLT-6739

**Comment by lfbrock** [comment]
@asaadmahmood - so I think the scroll issue is fixed, but if I put a message on that's longer than one line, it covers the UI again. If text is one line it works fine. 

Thoughts on fixing this? I guess the options would be to either truncate if the message is too long for the permanent banner, or let it go onto multiple lines and stop it from overlapping. I'm not sure what the easiest fix is - for the use case we have it wouldn't go onto multiple lines so it might be a rare case, but we should fix the corner case regardless. 

screenshot 2017-06-02 13 07 54

Also I think the on hover for the Site URL message also needs to be fixed, on previous version we just left it white on hover: 
screenshot 2017-06-02 13 03 26

Other than that everything else looks good

**Comment by jwilander** [comment]
@lfbrock I merged @asaadmahmood's fix for the above issue into this branch

**Comment by lfbrock** [comment]
@asaadmahmood - the Site URL on hover is still broken: 
screenshot 2017-06-05 09 30 22

**Pull request merged**